src/interfaces/simulation.ts

```diff
@@ -37,85 +60,40 @@ export class SimulationInterface<T extends ChainId> extends ServiceInterface<T>
-    const vaultContract =
-      zapProtocol === ZapProtocol.PICKLE
-        ? new PickleJarContract(toVault, signer)
-        : new YearnVaultContract(toVault, signer);
-
-    const underlyingToken = await vaultContract.token().catch(() => {
-      throw new EthersError("failed to fetch token", EthersError.FAIL_TOKEN_FETCH);
-    });
-    const isZapping = underlyingToken !== sellToken;
-    let forkId: string | undefined;
-    let simulateDeposit: (save: boolean) => Promise<TransactionOutcome>;
 
-    if (isZapping) {
-      if (!options.slippage) {
-        throw new SdkError("slippage needs to be specified for a zap", SdkError.NO_SLIPPAGE);
-      }
+    const vaultContract = this.getVaultContract({ toVault, signer });
 
-      let needsApproving: boolean;
+    const [vault] = await this.yearn.vaults.getDynamic([toVault]);
 
-      if (sellToken === EthAddress) {
-        needsApproving = false;
-      } else {
-        needsApproving = await this.yearn.services.zapper
-          .zapInApprovalState(from, sellToken, zapProtocol)
-          .then((state) => !state.isApproved)
-          .catch(() => {
-            throw new ZapperError("approval state", ZapperError.ZAP_IN_APPROVAL_STATE);
-          });
-      }
+    const token = await this.yearn.tokens.findByAddress(sellToken);
 
-      forkId = needsApproving ? await this.simulationExecutor.createFork() : undefined;
-      options.forkId = forkId;
-
-      const approvalTransactionId = needsApproving
-        ? await this.yearn.services.zapper
-            .zapInApprovalTransaction(from, sellToken, "0", zapProtocol)
-            .catch(() => {
-              throw new ZapperError("approval", ZapperError.ZAP_IN_APPROVAL);
-            })
-            .then(async (approvalTransaction) => {
-              return await this.simulateZapApprovalTransaction(approvalTransaction, options);
-            })
-            .then((res) => res.simulation.id)
-        : undefined;
-      options.root = approvalTransactionId;
+    const { isZapInSupported, zapInWith } = getZapInDetails({ chainId: this.chainId, token, vault });
 
-      simulateDeposit = (save: boolean): Promise<TransactionOutcome> => {
-        options.save = save;
-        return this.zapIn(
-          from,
-          sellToken,
-          underlyingToken,
-          amount,
-          toVault,
-          vaultContract,
-          zapProtocol,
-          needsApproving,
-          options
-        );
-      };
-    } else {
-      const needsApproving = await this.depositNeedsApproving(from, sellToken, toVault, amount, signer);
+    if (isZapInSupported && zapInWith) {
+      return this.handleZapInSimulationDeposit({ depositArgs, zapInWith, vaultContract });
+    }
 
-      forkId = needsApproving ? await this.simulationExecutor.createFork() : undefined;
-      options.forkId = forkId;
+    return this.handleDirectSimulationDeposit({ depositArgs, vaultContract, signer });
```

Comment:
Just to make sure users get an error when trying to deposit an invalid token, we should add a validation to throw an error when `!isZapInSupported && !isUnderlyingToken`. Right now all invalid token zaps will be treated as direct deposit.

Reply:
Will a direct deposit have one of those?

Reply:
Added the check with https://github.com/yearn/yearn-sdk/pull/280/commits/97bf27505bd6e115676caa8486579e91261cc878 -- so now only when depositing the underlying token we should go to direct deposits